libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/NavigationRouteProcessor.kt

```diff
@@ -0,0 +1,141 @@
+    private var routeGeometryWithBuffer: Geometry? = null
+
+    fun buildNewRouteProgress(navigator: MapboxNavigator, status: NavigationStatus, route: DirectionsRoute): RouteProgress? {
+        previousStatus = status
+        updateRoute(route, navigator)
+        return buildRouteProgressFrom(status, navigator)
+    }
+
+    fun updatePreviousRouteProgress(routeProgress: RouteProgress?) {
+        previousRouteProgress = routeProgress
+    }
+
+    fun retrievePreviousRouteProgress() =
+            previousRouteProgress
+
+    fun retrievePreviousStatus() =
+            previousStatus
+
+    private fun updateRoute(route: DirectionsRoute, navigator: MapboxNavigator) {
+        if (this.route == null || this.route != route) {
+            this.route = route
+            routeGeometryWithBuffer = navigator.retrieveRouteGeometryWithBuffer()
+        }
+    }
+
+    private fun buildRouteProgressFrom(status: NavigationStatus, navigator: MapboxNavigator): RouteProgress? {
+        val legIndex = status.legIndex
+        val stepIndex = status.stepIndex
+        val upcomingStepIndex = stepIndex + ONE_INDEX
+        route?.let { directionsRoute ->
+            updateSteps(directionsRoute, legIndex, stepIndex)
+            updateStepPoints(directionsRoute, legIndex, stepIndex, upcomingStepIndex)
+
+            val legDistanceRemaining = status.remainingLegDistance.toDouble()
+            val routeDistanceRemaining = NavigationHelper.routeDistanceRemaining(legDistanceRemaining, legIndex, directionsRoute)
+            val stepDistanceRemaining = status.remainingStepDistance.toDouble()
+            val legDurationRemaining = status.remainingLegDuration / ONE_SECOND_IN_MILLISECONDS
+
+            currentLegAnnotation = NavigationHelper.createCurrentAnnotation(currentLegAnnotation, currentLeg, legDistanceRemaining)
+            val routeState = status.routeState
+            val currentRouteState = progressStateMap[routeState]
+
+            val progressBuilder = RouteProgress.builder()
+                    .distanceRemaining(routeDistanceRemaining)
+                    .legDistanceRemaining(legDistanceRemaining)
+                    .legDurationRemaining(legDurationRemaining)
+                    .stepDistanceRemaining(stepDistanceRemaining)
+                    .directionsRoute(directionsRoute)
+                    .currentStep(currentStep)
+                    .currentStepPoints(currentStepPoints)
+                    .upcomingStepPoints(upcomingStepPoints)
+                    .stepIndex(stepIndex)
+                    .legIndex(legIndex)
+                    .inTunnel(status.inTunnel)
+                    .currentState(currentRouteState)
+
+            addRouteGeometries(progressBuilder)
+            addVoiceInstructions(status, progressBuilder)
+            addBannerInstructions(status, navigator, progressBuilder)
+            addUpcomingStepPoints(progressBuilder)
+            return progressBuilder.build()
+        }
+        return null
```

Comment:
See comment above: re: `RouteProgress?` should be non-nullable

Reply:
same as a bove